Ban parser: don't accept object name prefixes

walid-git opened the pull request:
Fixes #3962

walid-git commented:
Reworked the PR to also avoid bans on "req.ur".
Also added 4f17a68ce6406a92e7316eee0e07725709dcfa0d to prevent ban expressions with empty http header names that can lead to a panic (see [my comment](https://github.com/varnishcache/varnish-cache/issues/3962#issuecomment-1668222750) for details)

bsdphk commented:
Looks good.

nigoroll reviewed (approved):
Thank you @walid-git

Dridi commented on bin/varnishd/cache/cache_ban_build.c:
```diff
@@ -246,7 +246,8 @@ BAN_AddTest(struct ban_proto *bp,
 		return (bp->err);
 
 	for (pv = pvars; pv->name != NULL; pv++)
-		if (!strncmp(a1, pv->name, strlen(pv->name)))
+		if ((!(pv->flag & BANS_FLAG_HTTP) && !strcmp(a1, pv->name))
+			|| (pv->flag & BANS_FLAG_HTTP && !strncmp(a1, pv->name, strlen(pv->name))))
 			break;
```
Maybe decompose the loop to make it more readable?

```c
for (pv = pvars; pv->name != NULL; pv++) {
        is_http = pv->flag & BANS_FLAG_HTTP;
        if (!is_http && !strcmp())
            break;
        if (is_http && !strncmp())
            break;
}
```

Otherwise there's probably a line too long, the `||` should be on the previous line and this is the wrong indentation for the if continuation.

Also, this is lacking a test case.

Dridi commented on bin/varnishtest/tests/r03962.vtc:
```diff
@@ -0,0 +1,13 @@
+varnishtest "ban expression object name prefixes"
```
Please squash into first commit.

Dridi commented on bin/varnishd/cache/cache_ban_build.c:
```diff
@@ -258,6 +258,9 @@ BAN_AddTest(struct ban_proto *bp,
 
 	VSB_putc(bp->vsb, pv->tag);
 	if (pv->flag & BANS_FLAG_HTTP) {
+		if (strlen(a1 + strlen(pv->name)) < 1)
+			return (ban_error(bp,
+			    "An HTTP header name must be specified after \"%s\"", pv->name));
```
```c
"Missing header name: '%s'"
```

Dridi reviewed (changes requested):
A thought occurred to me, VCL supports "non symbolic" header names with the following syntax:

```vcl
req.http."valid.header.name"
```

But I don't think the ban parser learned about these.

walid-git commented:
Addressed @Dridi's comments. Also added changelog

> A thought occurred to me, VCL supports "non symbolic" header names with the following syntax:
> ```
> req.http."valid.header.name"
> ```
> But I don't think the ban parser learned about these.

This does't seem to be directly related to this issue, we should probably handle that separately
